sncosmo/specdata.py

```diff
@@ -0,0 +1,347 @@
+    bin_ends : `~numpy.ndarray`
+        The estimated end of each wavelength bin.
+    """
+    wave = np.asarray(wave)
+
+    # First order estimate
+    o1 = (wave[:-1] + wave[1:]) / 2.
+
+    # Second order correction
+    o2 = 1.5*o1[1:-1] - (o1[2:] + o1[:-2]) / 4.
+
+    # Estimate front and back edges
+    f2 = 2 * wave[1] - o2[0]
+    f1 = 2 * wave[0] - f2
+    b2 = 2 * wave[-2] - o2[-1]
+    b1 = 2 * wave[-1] - b2
+
+    # Stack everything together
+    bin_edges = np.hstack([f1, f2, o2, b2, b1])
+    bin_starts = bin_edges[:-1]
+    bin_ends = bin_edges[1:]
+
+    return bin_starts, bin_ends
+
+
+def _parse_wavelength_information(wave, bin_edges, bin_starts, bin_ends):
+    """Parse wavelength information and return a set of bin starts and ends
+
+    TODO: documentation
+    """
+    # Make sure that a valid combination of inputs was given.
+    valid_count = 0
+    if wave is not None:
+        valid_count += 1
+    if bin_edges is not None:
+        valid_count += 1
+    if bin_starts is not None and bin_ends is not None:
+        valid_count += 1
+    if valid_count != 1:
+        raise ValueError('must specify exactly one of wave, bin_edges, or '
+                         'bin_starts and bin_ends')
+
+    # Extract the bin starts and ends.
+    if wave is not None:
+        bin_starts, bin_ends = _estimate_bin_edges(np.asarray(wave))
+    elif bin_edges is not None:
+        bin_starts = bin_edges[:-1]
+        bin_ends = bin_edges[1:]
+    elif bin_starts is None or bin_ends is None:
+        raise ValueError('must specify both bin_starts and bin_ends')
+
+    # Make sure that the bin ends are larger than the bin starts.
+    if np.any(bin_starts >= bin_ends):
+        raise ValueError('bin_ends must be larger than bin_starts')
+
+    return bin_starts, bin_ends
+
+
+class SpectrumData(object):
+    """Standardized representation of spectroscopic data.
+
+    The edges of the spectral element wavelength bins are stored in
+    ``bin_edges`` as a numpy array. ``flux`` and ``fluxerr`` are the
+    corresponding flux values and uncertainties of the spectrum for each
+    spectral element, and are both numpy arrays with lengths of one less than
+    the size of ``bin_edges``.
+
+    Optionally, there is a single ``time`` associated with all of the
+    observations.
+
+    TODO: calibrated spectra? See spectrum.py where there is wave_unit and unit,
+    and everything is converted internally.
+    TODO: covariance?
+
+    Has attribute ``fluxcov`` which may be ``None``.
+
+    Parameters
+    ----------
+    TODO
+
+    """
+    def __init__(self, wave=None, flux=None, fluxerr=None, fluxcov=None,
+                 bin_edges=None, bin_starts=None, bin_ends=None, time=None):
+        # Extract the bin starts and ends
+        bin_starts, bin_ends = _parse_wavelength_information(
+            wave, bin_edges, bin_starts, bin_ends
+        )
+        self.bin_starts = bin_starts
+        self.bin_ends = bin_ends
```

Comment:
I'm curious about the benefit of allowing non-contiguous elements.

(If it were disallowed, I can see that `merge` would have to include a `rebin` operation).

Reply:
I think that handling non-contiguous spectra is outside of the scope of `sncosmo`, so I'm going to remove this. Things like fitting are a lot more complicated to optimize if you allow non-contiguous spectra.